connection_go18.go

```diff
@@ -0,0 +1,90 @@
+// +build go1.8
+
+// Go MySQL Driver - A MySQL-Driver for Go's database/sql package
+//
+// Copyright 2017 The Go-MySQL-Driver Authors. All rights reserved.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package mysql
+
+import (
+	"context"
+	"database/sql/driver"
+	"net"
+	"time"
+)
+
+func (mc *mysqlConn) Ping(ctx context.Context) error {
+	var err error
+
+	if ctx == nil {
+		return mc.pingImpl()
+	}
+	mc.ctx = ctx
+	defer func() {
+		mc.ctx = nil
+	}()
+
+	if deadline, ok := ctx.Deadline(); ok {
+		if err = mc.netConn.SetDeadline(deadline); err != nil {
+			return err
+		}
+	}
+
+	if ctx.Done() == nil {
```

Comment:
Would `ctx.Done()` return nil?
Example code in [this document](https://golang.org/pkg/context/#Context) doesn't check nil.